packages/flutter/lib/src/widgets/restoration_properties.dart

```diff
@@ -496,3 +496,113 @@ class RestorableTextEditingController extends RestorableChangeNotifier<TextEditi
     return value.text;
   }
 }
+
+/// A [RestorableProperty] that knows how to store and restore a nullable [Enum]
+/// type.
+///
+/// {@macro flutter.widgets.RestorableNum}
+///
+/// The values are serialized using the name of the enum, obtained using the
+/// [EnumName.name] extension accessor.
+///
+/// The represented value is accessible via the [value] getter.
+class RestorableEnumN<T extends Enum> extends RestorableValue<T?> {
+  /// Creates a [RestorableEnumN].
+  ///
+  /// {@macro flutter.widgets.RestorableNum.constructor}
+  RestorableEnumN(T? defaultValue, { required this.allowedValues })
+    : _defaultValue = defaultValue;
+
+  @override
+  T? createDefaultValue() => _defaultValue;
+  final T? _defaultValue;
+
+  /// {@macro flutter.widgets.RestorableEnum.allowedValues}
+  ///
+  /// In addition to this list, because this [RestorableProperty] allows null,
+  /// the set of allowed values will also include null.
+  Iterable<T> allowedValues;
+
+  @override
+  void didUpdateValue(T? oldValue) {
+    if (value != oldValue) {
+      notifyListeners();
+    }
+  }
+
+  @override
+  T? fromPrimitives(Object? data) {
+    if (data == null) {
+      return null;
+    }
+    if (data is String) {
+      for (final T allowed in allowedValues) {
+        if (allowed.name == data) {
+          return allowed;
+        }
+      }
+    }
+    return _defaultValue;
+  }
+
+  @override
+  Object? toPrimitives() => value?.name;
+}
+
+
+/// A [RestorableProperty] that knows how to store and restore an [Enum]
+/// type.
+///
+/// {@macro flutter.widgets.RestorableNum}
+///
+/// The values are serialized using the name of the enum, obtained using the
+/// [EnumName.name] extension accessor.
+///
+/// The represented value is accessible via the [value] getter.
+class RestorableEnum<T extends Enum> extends RestorableValue<T> {
+  /// Creates a [RestorableEnumN].
+  ///
+  /// {@macro flutter.widgets.RestorableNum.constructor}
+  RestorableEnum(T defaultValue, { required this.allowedValues })
+    : _defaultValue = defaultValue;
+
+  @override
+  T createDefaultValue() => _defaultValue;
+  final T _defaultValue;
+
+  /// {@template flutter.widgets.RestorableEnum.allowedValues}
+  /// The set of allowed values that this [RestorableEnumN] may represent.
+  ///
+  /// This is a required field that determines which enum values may be
+  /// serialized and restored.
+  ///
+  /// If a value is encountered that is not in this set, the default value
+  /// supplied to the constructor is substituted for it.
+  ///
+  /// It is typically set to the `values` list of the enum type, but may also be
+  /// a subset of those values.
+  /// {@endtemplate}
+  Iterable<T> allowedValues;
```

Comment:
Do we need this feature? It seems unnecessarily complex that for the default case (I want to store/restore all possible enum values) you have to do extra work to set allowedValues.

It's also somewhat confusing that it only restores certain values and others get silently replaced with the default. It feels like this feature shouldn't really be in the base class for restoring enums...

Reply:
(Also looks like this feature is missing a test should we decide to keep it)

Reply:
Makes sense. Let's document what one would typically pass in for this parameter (e.g. Enum.values) then. Maybe we should also reword this doc a little. Right now it sounds like its purpose is to filter certain values from serialization.

I would actually go one step further and disallow filtering altogether because it is kinda confusing. Maybe we should rename the param to something like "allValues" and then assert that a value you set on the RestorableEnum is actually part of the allValues set? Similarly, if a serialized value is received that's not part of allvalues, maybe we should just assert instead of silently dropping it?

Reply:
Yes, OK.

One thing I worry about, though, is what happens when someone renames an enum value: should it assert when it tries to read an old name from the serialized data and put it into a new one? Or does the restoration data get erased when an app is upgraded? In the current code, it'll just get a default value.  I suppose it's just an assert, and in release mode it would get a default value anyhow, but in that case does it make sense to assert?

I briefly considered adding an optional "conversion" function that would be able to filter the incoming values and convert them to "new" values if old values were encountered, but I thought that was probably overkill for enums, and there should probably be a more central version of that functionality. I didn't want to re-invent protos.

Reply:
If I remember correctly, the restoration data is thrown away when the app is upgraded.

Reply:
Oh, excellent, so we don't need to worry about that, really.  OK, I added the asserts and tests.